Alternative implementation of ThreadPoolWait.

jinlmsft opened the pull request:
ThreadPoolWait implement functionality of ThreadPoolWait .RegisterWaitForSingleObject, with three capability:
1. To remove a particular wait handle and cancel its operation
2. To show all wait handles queued in the system,
   and 3. To show long continuation (not normal).

The old implementation uses a customized threadpool. The new implementation uses ThreadPoolWait.RegisterWaitForSingleObject.

One potential performance issue is that:

ThreadPoolWait.MaxSpinMs = 200 ms.

This is the time between the callback is called, and the registeredWaitHandle is put into the system. I found that if I set

ThreadPoolWait.MaxSpinMs = 50 ms.

Sometime, the system will show it is unable to grab the registeredWaitHandle in 50ms. I can't reason why it takes so long for the
ThreadPoolWait.RegisterWaitForSingleObject to get back. In slower system, we may need to use a even larger wait.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle) |> ignore
+                    rwh.Unregister(null) |> ignore 
+                else
+                    failwith (sprintf "ThreadPoolWait.Callback, can't find the entry of waithandle and continuation object after %d ms, reason: %s " ThreadPoolWait.MaxSpinMs !failedReason )
```
This is unrecoverable error. On slow system, it can totally happen. However, there's nothing wrong about it, I think we should not fail here. 

I think if we handle "rwh" as the example in the email, then we can add the item to the handleCollection before calling RegisterWaitForSingleObject.  As a result, there is no need to spin here, it will be a deterministic behavior. 

BTW: I assume the purpose WaitHandleCollection is to help on monitoring and remove?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+            for pair in handleCollection do 
+                let rwh, infoFunc, workRef, refStartContinuation = pair.Value
+                // If this get executed, the actual work will be cancelled. 
+                Volatile.Write( workRef,  (( fun _ -> () ), null) )  
```
This code seems rather unusual way to achieve cancellation. Also it can be dangerous, for example, at

   let continuation, unblockHandle = !workRef

depends on the F# implementation,  we may get continuation before it was set to null, but "unblockHandle" after it becomes null?  It's totally OK for the compiler to generate code like

   continuation = workRef.Item1
   unblockHandle = workRef.Item2

I'd recommend not doing this. 

Let's take a step back: In what case, we would want to remove a handle without getting cancellation request?  I think the "callback" should take a CancellationToken as a input parameter, and do cooperative cancellation inside it.  We also put the cancellation into the collection. And we should further assume TryRemove will be called only if the cancellation is requested.   Here, we can check whether the cancellation has been requested. If not,w e should regard it as an error and do not try to remove the handle. 

If we really really have the case where we want to remove the WaitHandle but somehow has no CancellationToken from upstream (I really doubt that should be the case),  we can try to create a CancellationTokenSource and add it to collection too (should be null, if we already have a token from upstream).  In this case, we get a token from our own just created source and pass it to then callback. During TryRemove, if we find CancellationTokenSource is not null, then we request the cancel, then set the handle.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (handle,continuationObject) , -1, true )
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(continuationObject) <- ( rwh, infoFunc, ref (continuation, unblockHandle), ref DateTime.MinValue.Ticks )
+    static member private CallBack( state: Object ) (timeout:bool) = 
```
the callback should take cancellation token as a input parameter, so it can do cooperative cancellation.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+                Logger.LogF( LogLevel.MildVerbose, ( fun _ -> sprintf "ThreadPoolWait, threadpool %s is the last registered threadpools, try terminating all" name ))
+            // Termniate all Wait thread. 
+            ThreadPoolWait.TryTerminateAll()
+    static member WaitForHandle (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) =
```
should take in a cancellation token.

zhuweirong replied:
user should also provide a cancellation token (which can be null), thus we can do cooperative cancellation

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+        else
+            let continuationObject = continuation :> Object
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (handle,continuationObject) , -1, true )
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
```
the cancellation token should be put into the collection too, thus can be used later (see comments at TryRemove)

zhuweirong commented:
Great stuff! I have some comments in the code, please check them out.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static let rwhToRemove = ConcurrentQueue<_>()
+    static let rwhLookup = ConcurrentDictionary<_,_>()
+    /// Maximum wait for the RegisteredWaitHandle to show up
+    static member val private MaxSpinMs = 0 with get, set
```
MaxSpinMs seems not used anywhere, should be removed.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static let rwhLookup = ConcurrentDictionary<_,_>()
+    /// Maximum wait for the RegisteredWaitHandle to show up
+    static member val private MaxSpinMs = 0 with get, set
+    static member val WarningWaitInMs = 10 with get, set
```
Not used either.

jinlmsft replied:
Will remove

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static member val WarningWaitInMs = 10 with get, set
+    static member val ActiveThreadPools = ConcurrentDictionary<_,_>() with get
+    static member val WaitHandleCollection = ConcurrentDictionary<_,ConcurrentDictionary<_,_>>() with get
+    static member RegisterThreadPool (name:string) = 
```
Should ActiveThreadPools/WaitHandleCollections be private?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static member val private MaxSpinMs = 0 with get, set
+    static member val WarningWaitInMs = 10 with get, set
+    static member val ActiveThreadPools = ConcurrentDictionary<_,_>() with get
+    static member val WaitHandleCollection = ConcurrentDictionary<_,ConcurrentDictionary<_,_>>() with get
```
Should ActiveThreadPools/WaitHandleCollections be private?

jinlmsft replied:
will change to static let.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+        Logger.LogF( LogLevel.MildVerbose, ( fun _ -> sprintf "ThreadPoolWait, unregister wait pool %s ............" name ))
+        ThreadPoolWait.ActiveThreadPools.TryRemove( name ) |> ignore
+        if ThreadPoolWait.ActiveThreadPools.IsEmpty then 
+            if ( !(ThreadPoolWaitDeprecated.NumberOfWaitingThreads) > 0 ) then 
```
Why check ThreadPoolWaitDeprecated?

jinlmsft replied:
will remove RegisterThreadPool and UnregisterThreadPool.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+        Logger.LogF( LogLevel.MediumVerbose, ( fun _ -> sprintf "ThreadPoolWait, register wait pool %s ............" name ))
+        ThreadPoolWait.ActiveThreadPools.Item( name ) <- DateTime.UtcNow
+    static member UnregisterThreadPool (name:string) = 
+        Logger.LogF( LogLevel.MildVerbose, ( fun _ -> sprintf "ThreadPoolWait, unregister wait pool %s ............" name ))
```
When should RegisterThreadPool/UnregisterThreadPool be called? What's the purpose? Or they are here for the purpose of backwards compatibility for existing code? Either way, please add some comments?

jinlmsft replied:
The original thought is to track a pool of handles related to a particular job. But with the job lifecycle object, it may not be necessary, as we can simply register handle wakeup through the CancellationToken with related to the job. I decide to remove both RegisterThreadPool  and UnregisterThreadPool.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+    /// Maximum wait for the RegisteredWaitHandle to show up
+    static member val private MaxSpinMs = 0 with get, set
+    static member val WarningWaitInMs = 10 with get, set
+    static member val ActiveThreadPools = ConcurrentDictionary<_,_>() with get
```
What's the purpose of ActiveThreadPools?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+            let jobObject = Object()
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (infoFunc,handle,continuation,unblockHandle,jobObject) , -1, true )
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(jobObject) <- ( rwh, infoFunc, ref DateTime.MinValue.Ticks )
```
so we assume that more than one WaitCallback can be registered to wait on the same handle, right?

jinlmsft replied:
that is correct. Multiple waitcallback can be registered to wait on a handle.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+                    rwh.Unregister(null) |> ignore 
+                    handleCollection.TryRemove( jobObject ) |> ignore 
+                    if handleCollection.IsEmpty then
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle ) |> ignore 
```
This is not safe to do. It's possible this thread sees handleCollection.IsEmpty, then takes the branch, however, before it called "TryRemove", another thread can register a wait for the same handle.  This remove would cause the new handle to be removed, which should not happen.

jinlmsft replied:
Will redesign. Use only jobObject as a single level ConcurrentDictionary, instead of two level.

zhuweirong commented:
In the test run https://ci.appveyor.com/project/zhuweirong/prajna-rbpft/build/0.0.1.259, though it's successful, it seems the test time has significantly increased (compared to latest merged pull request: https://ci.appveyor.com/project/zhuweirong/prajna-rbpft/build/0.0.1.254). (I'm running it again to see if it's just a glitch).

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+                    let bExist, tuple = handleCollection.TryGetValue(jobObject)  
+                    if bExist then 
+                        let rwh, _, _ = tuple       
+                        rwh.Unregister(null) |> ignore 
```
On this code path, do you also need to remove the jobObject from the handleCollection, and remove the handleCollection from WaitHandleCollection?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+                Volatile.Write( ticksRef, DateTime.MaxValue.Ticks )
+        match handle with 
+        | :? ManualResetEvent as ev -> 
+            ev.Set() |> ignore 
```
based on the code here, it seems we will set the event even when bExits = false,  why will this happen?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1814,16 +1951,17 @@ and [<AllowNullLiteral>]
+//                                    let elapseMs = ( cur - continueTicks )/ TimeSpan.TicksPerMillisecond
+//                                    Logger.LogF( LogLevel.Info, ( fun _ -> sprintf "Possible Deadlocks! cont() of %s has been in executionin ThreadPoolWait for more than %d ms. Possible deadlock or poor performance in implementing cont() (there should not be any blocking operation or long running operation in cont()) " 
+//                                                                                       (pool.ContinueInfo())
+//                                                                                       elapseMs ))
```
just remove the dead code?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+                reraise()
+        else
+            failwith "ThreadPoolWait.Callback, timeout path reached (impossible execution route)"
+    static member ToUnregister() = 
```
should be private

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+            for pair in handleCollection do 
+                let rwh, _, ticksRef = pair.Value
+                // Try to marked job as cancelled, 
+                Volatile.Write( ticksRef, DateTime.MaxValue.Ticks )
```
Just to call out that this is another place that we rely on 64-bit read/write being atomic.

zhuweirong commented:
The 2nd run failed: https://ci.appveyor.com/project/zhuweirong/prajna-rbpft/build/0.0.1.260, I'm re-running it again.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static let rwhLookup = ConcurrentDictionary<_,_>()
+    static let waitHandleCollection = ConcurrentDictionary<_,_>() 
+    static let unregisterTimer = new System.Threading.Timer( TimerCallback( ThreadPoolWait.ToUnregister ), null, 10, 10 )
+    static member WaitForHandleWithCancellation (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) (token:CancellationToken) =
```
the "token" is not used. Is it for future use?  please add a comment to avoid the confusion here?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static let waitHandleCollection = ConcurrentDictionary<_,_>() 
+    static let unregisterTimer = new System.Threading.Timer( TimerCallback( ThreadPoolWait.ToUnregister ), null, 10, 10 )
+    static member WaitForHandleWithCancellation (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) (token:CancellationToken) =
+        token.Register( fun _ -> ThreadPoolWait.TryWakeup handle |> ignore ) |> ignore 
```
should we first check if the token has already received the cancellation request?

zhuweirong replied:
(Even with the check I mentioned above) it is possible that we register the "TryWakeUp" with the token,  but a cancellation request happens before we move to the WaitForHandle (or in mid of it), the TryWakeUp is called and did nothing, and we go ahead with WaitForHandle.   The point is:  register TryWakeUp with token does not ensure it will indeed do the wakeup.  So we cannot completely rely on it.  Do we have any other measures? If we do, is this  "register" still useful?

# 

I just realized that TryWakeUp is only to signal the handle to wait, it should be fine then ...

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+                    let rwh, _, _, _ = tuple
+                    rwh.Unregister(null) |> ignore 
+                else
+                    requeue.Add( jobObject )
```
in what scenario that it is possible that the jobObjects exists in rwhToRemove but not in waitHandleCollection?  If happens, should this be an error that we should log and investigate?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+                else
+                    requeue.Add( jobObject )
+        for jobObject in requeue do 
+            rwhToRemove.Enqueue()    
```
hmm, the Enqueue does not need to take any parameter?

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+/// and 3. To show long continuation (not normal). 
+type internal ThreadPoolWait() = 
+    static let rwhToRemove = ConcurrentQueue<_>()
+    static let rwhLookup = ConcurrentDictionary<_,_>()
```
rwhLookup is not used anywhere.

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+            let _, _, _, handle = pair.Value
+            ThreadPoolWait.TryWakeup( handle ) |> ignore 
+        ()
+    /// Mark all functions associated with handle as cancelled, the continuation function will not execute
```
The comment is not accurate.  The continuation function will be executed, we are depending on the caller to cancel them properly before calling this function.  (BTW: this is the semantics I prefer).

zhuweirong commented on src/tools/tools/process.fs:
```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+            ev.Set() |> ignore 
+            true
+        | _ -> 
+            false
```
Let's log here so we can identify such use cases.

jinlmsft committed:
Alternative implementation of ThreadPoolWait.

ThreadPoolWait implement functionality of ThreadPoolWait .RegisterWaitForSingleObject, with three capability:

1. To wakeup a particular wait handle ( and cancel its operation, may be removed later, as cancellation should be implemented within continuation)
2. To show all wait handles queued in the system,
and 3. To show long continuation (not normal).

The old implementation uses a customized threadpool. The new implementation uses ThreadPoolWait.RegisterWaitForSingleObject.
